CDAP-10600 complete dynamic partitioner plugin

caskchester opened the pull request:
https://issues.cask.co/browse/CDAP-10600

1. Added support for avro output format.
2. Added unit tests for pipeline configuration.
3. Explore queries now load in UI.
4. Added jump links in explore window.
5. Writing to same partition the second time fails is a CDAP issue because it doesn't check whether a partition exists before creating a new partition. It treats all incoming partitions as new partitions. Therefore, a dynamic partitioning pipeline would fail if it attempts to write to an existing partition.

russorat commented on pom.xml:
```diff
@@ -264,6 +264,17 @@
     </dependency>
+    <dependency>
+      <groupId>org.apache.orc</groupId>
+      <artifactId>orc-mapreduce</artifactId>
```
do you need this even though you aren't writing in the orc format?

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java:
```diff
@@ -0,0 +1,215 @@
+/**
+ * A {@link BatchSink} to write Avro records to a {@link PartitionedFileSet}.
+ */
+@Plugin(type = "batchsink")
```
BatchSink.PLUGIN_TYPE

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java:
```diff
@@ -0,0 +1,215 @@
+ * A {@link BatchSink} to write Avro records to a {@link PartitionedFileSet}.
+ */
+@Plugin(type = "batchsink")
+@Name("DynamicPFSAvro")
```
Let's make the name longer and more readable. "DynamicPartitionedAvroFile" should be good. Class names and stuff can stay the same.

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java:
```diff
@@ -0,0 +1,215 @@
+    String basePath = config.basePath == null ? pfsName : config.basePath;
+
+    // parse to make sure it's valid
+    new org.apache.avro.Schema.Parser().parse(config.schema);
```
Can you move all the validation logic to a separate validate function in the config class to match many of the other plugins? Also, what happens if the list of partition fields is macro enabled? You will need to have a validation function that checks for all this stuff at configure time and also during the initialize when the runtime args have been substituted.

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java:
```diff
@@ -0,0 +1,215 @@
+      parsedSchema = Schema.parseJson(config.schema);
+      hiveSchema = HiveSchemaConverter.toHiveSchema(parsedSchema);
+    } catch (UnsupportedTypeException | IOException e) {
+      throw new RuntimeException("Error: Schema is not valid ", e);
```
this should be an IllegalArgException.

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java:
```diff
@@ -0,0 +1,215 @@
+    PartitionedFileSetProperties.Builder properties = PartitionedFileSetProperties.builder();
+    addPartitionedFileSetProperties(properties);
+
+    pipelineConfigurer.createDataset(pfsName, PartitionedFileSet.class.getName(),
```
The dataset name should be macro enabled similar to the other sink plugins.

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetParquetSink.java:
```diff
@@ -105,7 +101,13 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
-      partitionBuilder.addStringField(partitionFields[i]);
+      // throw exception if the field used to partition does not exist
+      if(parsedSchema.getField(partitionFields[i]) == null) {
+        throw new IllegalArgumentException(String.format("Field %s is not a member of the input schema.",
```
The wording here is strange. "Not a member"? Maybe "Field %s is not present in the input schema..."

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java:
```diff
@@ -0,0 +1,215 @@
+  @Override
+  public void transform(StructuredRecord input,
+                        Emitter<KeyValue<AvroKey<GenericRecord>, NullWritable>> emitter) throws Exception {
+    emitter.emit(new KeyValue<>(new AvroKey<>(recordTransformer.transform(input)), NullWritable.get()));
```
what happens if there is a null value for one of the partition values?

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java:
```diff
@@ -0,0 +1,215 @@
+      for (int i = 0; i < fieldNames.length; i++) {
+        if (key.datum().get(fieldNames[i]) == null) {
+          // This call will result in an exception but there's nothing else I can do. [CDAP-7053]
+          keyBuilder.addStringField(fieldNames[i], null);
```
to avoid this, maybe there should be a check to make sure the partition columns are marked as "not nullable"

russorat commented on src/main/java/co/cask/hydrator/plugin/common/FileSetUtil.java:
```diff
@@ -28,6 +28,8 @@
 import org.apache.avro.mapreduce.AvroKeyOutputFormat;
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.mapreduce.Job;
+import org.apache.orc.mapreduce.OrcInputFormat;
```
remove since there is no orc support

russorat commented on src/test/java/co/cask/hydrator/plugin/DynamicPartitionFileSetSinkTest.java:
```diff
@@ -0,0 +1,115 @@
+    }
+  }
+
+  private void configureAvroSinkfromConfigs (String sinkName, Schema schema, String fieldnames,
```
remove space after function name

russorat commented on src/test/java/co/cask/hydrator/plugin/DynamicPartitionFileSetSinkTest.java:
```diff
@@ -0,0 +1,115 @@
+    dynamicPartitionFileSetAvroSink.configurePipeline(mockPipelineConfigurer);
+  }
+
+  private void configureParquetSinkfromConfigs (String sinkName, Schema schema, String fieldnames,
```
same

russorat commented on widgets/DynamicPFSAvro-batchsink.json:
```diff
@@ -0,0 +1,62 @@
+{
+  "metadata": {
+    "spec-version": "1.3"
```
i think we are at 1.4 but you should confirm

caskchester replied:
The UI people told me to use 1.3

caskchester commented:
1. Added macro support for dataset name, schema, and base path.
2. Restructured code.

russorat commented on docs/DynamicPartitionedAvroFileSet-batchsink.md:
```diff
@@ -0,0 +1,56 @@
+
+Properties
+----------
+**name:** Name of the ``PartitionedFileSet`` to which records are written.
```
Could you convert these properties into this table format? here and in the docs page.

russorat replied:
https://github.com/hydrator/decompress-action#properties

russorat commented on docs/DynamicPartitionedORCFileSet-batchsink.md:
```diff
@@ -0,0 +1,56 @@
+
+Properties
+----------
+**name:** Name of the ``PartitionedFileSet`` to which records are written.
```
convert to table format: https://github.com/hydrator/decompress-action#properties

russorat commented on docs/DynamicPartitionedORCFileSet-batchsink.md:
```diff
@@ -0,0 +1,56 @@
+``'create_date'`` of the record as the partition:
+
+    {
+        "name": "DynamicPartitionedORCFileSet",
```
remove json from this and all other docs and readme pages

caskchester replied:
So do we remove the whole Example. Or should we just replace the json with something else?

russorat commented on pom.xml:
```diff
@@ -398,6 +414,31 @@
+              org.apache.avro.mapreduce;
+              parquet.avro.*;
+              org.apache.orc.*;
+              <!-- Used by S3 source/sink -->
```
what s3 source/sink? is this needed?

caskchester replied:
not needed, removed

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java:
```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
2017

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java:
```diff
@@ -0,0 +1,164 @@
+ * A {@link BatchSink} to write Avro records to a {@link PartitionedFileSet}.
+ */
+@Plugin(type = BatchSink.PLUGIN_TYPE)
+@Name("DynamicPartitionedAvroFileSet")
```
Check with @bdmogal since we might be changing these names.

caskchester replied:
Avro Dynamic Partitioned Dataset.

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java:
```diff
@@ -0,0 +1,164 @@
+  }
+
+  @Override
+  public void prepareRun(BatchSinkContext context) throws DatasetManagementException{
```
space before {

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java:
```diff
@@ -0,0 +1,164 @@
+
+    @Override
+    public void initialize(MapReduceTaskContext mapReduceTaskContext) {
+      if (mapReduceTaskContext.getPluginProperties(DynamicPartitionFileSetAvroSink.NAME) == null) {
```
is this really needed? how was the pipeline published if the plugin doesn't exist?

caskchester replied:
removed

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetORCSink.java:
```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
2017

russorat commented on src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetORCSink.java:
```diff
@@ -0,0 +1,215 @@
+  }
+
+  @Override
+  public void prepareRun(BatchSinkContext context) throws DatasetManagementException{
```
space before {

russorat commented on docs/DynamicPartitionedORCFileSet-batchsink.md:
```diff
@@ -0,0 +1,56 @@
+**basePath:** Base path for the ``PartitionedFileSet``. Defaults to the name of the dataset.
+
+**compressionCodec:** Optional parameter to determine the compression codec to use on the resulting data. 
+Valid values are None, Snappy, GZip, and LZO.
```
does orc support all of these codecs?

russorat commented on src/main/java/co/cask/hydrator/plugin/common/FileSetUtil.java:
```diff
@@ -172,6 +140,40 @@ public static void configureAvroFileSet(String configuredSchema, FileSetProperti
+   * @param properties a builder for the file set properties
+   */
+  public static void configureORCFileSet(String configuredSchema, FileSetProperties.Builder properties)  {
+    //TODO test if complex cases run with lowercase schema only
```
is this TODO still valid?

caskchester replied:
I'll create a JIRA for this since this code is also used in the TPFS ORC plugin.

russorat commented on src/main/java/co/cask/hydrator/plugin/common/MacroParser.java:
```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
2017

russorat commented on src/main/java/co/cask/hydrator/plugin/common/MacroParser.java:
```diff
@@ -0,0 +1,218 @@
+/*
```
@bdmogal / @sreevatsanraman shouldn't this just be a class in CDAP and included? why does he need to include the file?

caskchester replied:
This code originally resides in app-fabric. It is better not to include app-fabric code in a plugin.

russorat commented on src/main/java/co/cask/hydrator/plugin/common/StructuredToOrcTransformer.java:
```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
2017

russorat commented on src/main/java/co/cask/hydrator/plugin/common/StructuredToOrcTransformer.java:
```diff
@@ -0,0 +1,128 @@
+        WritableComparable writable = convertToWritable(field, input);
+        orcRecord.setFieldValue(fields.get(i).getName(), writable);
+      } catch (UnsupportedTypeException e) {
+        LOG.debug("{} is not a supported type", field.getName(), e);
```
is this debug needed?

caskchester replied:
removed

russorat commented on src/test/java/co/cask/hydrator/plugin/DynamicPartitionFileSetSinkTest.java:
```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
2017

russorat commented on widgets/DynamicPartitionedAvroFileSet-batchsink.json:
```diff
@@ -0,0 +1,62 @@
+        {
+          "widget-type": "textbox",
+          "label": "Dataset Base Path",
+          "name": "basePath"
```
would it help to have placeholder text like the one used here? https://github.com/hydrator/decompress-action/blob/release/1.0/widgets/Decompress-action.json#L14

caskchester replied:
added

russorat commented:
Can you make sure this is noted on the jira about macro parsing so we know
to clean this up when it is fixed?

On Mon, Jun 26, 2017 at 11:10 AM, caskchester <notifications@github.com>
wrote:

> *@caskchester* commented on this pull request.
> ------------------------------
>
> In src/main/java/co/cask/hydrator/plugin/common/MacroParser.java
> <https://github.com/hydrator/dynamic-partitioner/pull/1#discussion_r124079837>
> :
>
> > @@ -0,0 +1,218 @@
> +/*
>
> This code originally resides in app-fabric. It is better not to include
> app-fabric code in a plugin.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hydrator/dynamic-partitioner/pull/1#discussion_r124079837>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AElVbX25X2f-DRQt918RzIEh9n-GdrfVks5sH_QAgaJpZM4N9CGu>
> .
>

caskchester commented:
addressed comments from @russorat and @bdmogal

russorat commented:
Replace it with an actual example. So show the data coming into the plugin,
then show the dataset (indicating the partitions) that the sink created.

On Mon, Jun 26, 2017 at 11:54 AM, caskchester <notifications@github.com>
wrote:

> *@caskchester* commented on this pull request.
> ------------------------------
>
> In docs/DynamicPartitionedORCFileSet-batchsink.md
> <https://github.com/hydrator/dynamic-partitioner/pull/1#discussion_r124090885>
> :
>
> > +If it doesn't exist, it will be created.
> +
> +**schema:** The Avro schema of the record being written to the sink as a JSON Object.
> +
> +**basePath:** Base path for the ``PartitionedFileSet``. Defaults to the name of the dataset.
> +
> +**compressionCodec:** Optional parameter to determine the compression codec to use on the resulting data.
> +Valid values are None, Snappy, GZip, and LZO.
> +
> +Example
> +-------
> +This example will write to a ``PartitionedFileSet`` named ``'users'`` using the original
> +``'create_date'`` of the record as the partition:
> +
> +    {
> +        "name": "DynamicPartitionedORCFileSet",
>
> So do we remove the whole Example. Or should we just replace the json with
> something else?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hydrator/dynamic-partitioner/pull/1#discussion_r124090885>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AElVbQIpFFX3cDNcigR3wOxidxo88Lllks5sH_6CgaJpZM4N9CGu>
> .
>

caskchester commented:
@russorat There are currently 2 somewhat minor issues related to the CDAP explore UI that I don't know how to fix. 
https://issues.cask.co/browse/CDAP-12007

Updated the docs file.

caskchester commented:
@russorat 
I don't know if using screenshots in the md files is a good idea. It shows up fine on github, but they don't resize properly in the cdap UI.

Renamed class and method names.

Please review

<img width="1280" alt="screen shot 2017-06-28 at 11 51 32 am" src="https://user-images.githubusercontent.com/28714384/27654874-3ef85996-5bf8-11e7-83dc-ef0b8c8f4392.png">

caskchester commented:
removed screenshots in the markdown files. This looks better IMO

russorat commented on docs/AvroDynamicPartitionedDataset-batchsink.md:
```diff
@@ -0,0 +1,69 @@
+----------
+| Configuration | Required | Default | Description |
+| :------------ | :------: | :------ | :---------- |
+| **name** | **Y** | None| Name of the ``PartitionedFileSet`` to which records are written. If it doesn't exist, it will be created. |
```
Can you make these names the same as what shows up in the UI for the plugin? Base Path instead of basePath for example?

russorat commented on docs/AvroDynamicPartitionedDataset-batchsink.md:
```diff
@@ -0,0 +1,69 @@
+| :------------ | :------: | :------ | :---------- |
+| **name** | **Y** | None| Name of the ``PartitionedFileSet`` to which records are written. If it doesn't exist, it will be created. |
+| **schema** | **Y** | None | The Avro schema of the record being written to the sink as a JSON Object. |
+| **basePath** | **N** | [Namepsace]/data/[Dataset name] | Base path for the ``PartitionedFileSet``. Defaults to the name of the dataset. |
```
Namespace is misspelled

russorat commented on docs/ORCDynamicPartitionedDataset-batchsink.md:
```diff
@@ -0,0 +1,73 @@
+| :------------ | :------: | :------ | :---------- |
+| **name** | **Y** | None| Name of the ``PartitionedFileSet`` to which records are written. If it doesn't exist, it will be created. |
+| **schema** | **Y** | None | The Avro schema of the record being written to the sink as a JSON Object. |
+| **basePath** | **N** | [Namepsace]/data/[Dataset name] | Base path for the ``PartitionedFileSet``. Defaults to the name of the dataset. |
```
namespace is misspelled

russorat commented:
Minor spelling mistake and updates to docs, otherwise LGTM

caskchester commented:
@russorat 
Fixed wording in docs.

Commit:
Add Avro and ORC output formats.
Resturctured code.
Enabled Macros.
Performs configuration checks during configure time.
Added unit tests.
